### ISSUE TITLE
Empty Dataframe

### ISSUE THREAD

**Issue by jonathanng** [ranaroussi/fix-yahoo-finance#22]
I'm getting empty dataframes.

```
from pandas_datareader import data as pdr
import fix_yahoo_finance
import datetime

df = pdr.get_data_yahoo(
    tickers = 'COT',
    start   = datetime.date(2017, 6, 4),
    end     = datetime.date(2017, 6, 21)
)

df
```

But there is data on Yahoo:

https://finance.yahoo.com/quote/COT/history?p=COT

**Comment by ranaroussi** [comment]
Try upgrading to v 0.0.18 and try again

Upgrade using:

```bash
$ pip install fix_yahoo_finance --upgrade --no-cache-dir
```

**Comment by maximus12793** [comment]
I am getting the same issue(s) today. SPY/BABA/QIHU etc. some work but the vast majority break

**Comment by renium** [comment]
I updated to version 0.0.18.
Still approximately every third request returns an empty dataframe only.
I really doubt, that this is a Yahoo error.

**Comment by AnthonyFJGarner** [comment]
Hum...after about the third request every data frame is empty....

**Comment by maximus12793** [comment]
yes^. as a temp fix I wrapped any calls with the retry package which seems to generally work although am usually throttled after 20-30 requests.

**Comment by AnthonyFJGarner** [comment]
Oh blimey a valiant effort but that makes life very inconvenient.....

**Comment by maximus12793** [comment]
Logically the code makes sense I really feel as if yahoo may refresh tokens or blacklist requests quickly? I changed the code a bit to get a new token per response vs per session as it was written. That seems to work for a bit after which it breaks/is throttled

**Comment by AnthonyFJGarner** [comment]
Quandl US stock data is pretty cheap. Trouble is if course yahoo also provides FX, indices and mutual funds. And its coverage is worldwide. But I guess I may have to swallow the cost and buy it all.

**Comment by AnthonyFJGarner** [comment]
Gentlemen, god knows why but my downloads are now working a dream. No empty frames. Sample FX code attached in case it is of any use to anybody. A big thanks to ranourussi for this fix. Hope it isn't TOO temporary

`def fix_yahoo_FX(ticker_list, target_directory, start=None, end=None, source='yahoo',
              forward_fill=True):
    f = open(ticker_list)
    csv_f = csv.reader(f)
    next(csv_f)
    if start is None:
        start = datetime.date(1980, 1, 1)
    if end is None:
        end = datetime.date.today()

    for row in csv_f:
        try:
            ticker = row[3]
            direction = int(row[4])
            col_name = row[1]
            filename = (target_directory + col_name + '.csv')
            if not os.path.isfile(filename):
                df = yf.download(ticker, start, end)
                if forward_fill is True:
                    df = df.fillna(method='ffill')

                cols = [0, 1, 2, 3, 5]
                df.drop(df.columns[cols], axis=1, inplace=True)
                df.columns = [col_name]
                df.index.name = 'Date'
                df.to_csv(filename)
            else:
                old_file = pd.read_csv(filename, parse_dates=True, index_col=0)
                newstart = (pd.Series(old_file.index).iloc[-1]).to_datetime()
                df = yf.download(ticker, newstart, end)
                cols = [0, 1, 2, 3, 5]
                df.drop(df.columns[cols], axis=1, inplace=True)
                if forward_fill is True:
                    df = df.fillna(method='ffill')
                else:
                    pass
                df.columns = [col_name]
                df.index.name = 'Date'
                df.drop(df.head(1).index, inplace=True)
                df.to_csv(filename, mode='a', header=False)
        except CParserError:
            logger.fatal('Failed to update' + " " + ticker)
            logger.fatal(traceback.format_exc())
            pass
    f.close()`

**Comment by yoshiserry** [comment]
i'm still getting empty data frames too regardless of if I try to get data for only 5 to 10 tickers for 5-10 years. `mydata = yf.download(count[0:11], start="2000-01-01", end="2017-07-27")`

**Comment by renium** [comment]
Same here again. I extented the piece of code, listed above, and tested it with Python 2 and Python 3.6 .
Result: Lots of empty dataframes. The first ticker was ok, next tickers delivered always empty dataframes.
I am sorry, but the fix_yahoo_finance package is, in my eyes, completely unreliable. 
And I doubt, that Yahoo is the problem, because in my Wealthlab-Software the Yahoo-data-provider has been working again for weeks now.
In comparison, the tiingo- and quandl- API do work flawlessly with my code.
Drawback: Accessing more than 500 tickers costs money with tiingo.

**Comment by maximus12793** [comment]
Honestly just pay 5/mo or whatever it is and use tiingo api it is much faster and way more reliable in my experience!

Sent from Mail for Windows 10

**Comment by IanWorthington** [comment]
tiingo doesn't, afaik, have the UK data.

The first few runs today of yf.download returned empty frames, but after that it started working again.

(I always pickle my returned data to avoid hitting yahoo too hard.)

**Comment by SungchulLee** [comment]
I downloaded sp500 stock data from yahoo using pdr.get_data_yahoo. Many successes, but some empty dataframes. Here are tickers with empty dataframes, but not all:

(ticker != 'BHGE') and (ticker != 'SRCL') and (ticker != 'SYK') and (ticker != 'STI') and (ticker != 'SYMC') and (ticker != 'SYF') and (ticker != 'SNPS') and (ticker != 'SYY') and (ticker != 'TROW') and (ticker != 'TGT') and (ticker != 'TEL') and (ticker != 'FTI') and (ticker != 'TXN') and (ticker != 'TXT') and (ticker != 'UA')

Please help.

**Comment by yoshiserry** [comment]
@SungchuLee I think I've experienced a similar issue to what you have described I've been able to get most of the data just using short bursts of 100-200 stocks at a time. My time frame has been the last 30 years and only some failed. The issue may be not allowing enough time between your requests as well as the time frame you try and grab data for.

**Comment by SungchulLee** [comment]
@yoshiserry Following your advice, I added a line "time.sleep(1)" between the downloads and I've been able to get most of the data. Many thanks. Still some failed for new-born babies.

**Comment by yoshiserry** [comment]
I experimented with times after finding one second too short. I used what other people reported on this issue between 45 or 60 seconds

**Comment by SungchulLee** [comment]
@yoshiserry Following your advice, I added a line "time.sleep(60)" between the downloads and I've been able to get most of the data. Many thanks. Still some fails.

Previous failed list (not all) without time.sleep:
(ticker = 'BHGE') and (ticker = 'SRCL') and (ticker = 'SYK') and (ticker = 'STI') and (ticker = 'SYMC') and (ticker = 'SYF') and (ticker = 'SNPS') and (ticker = 'SYY') and (ticker = 'TROW') and (ticker = 'TGT') and (ticker = 'TEL') and (ticker = 'FTI') and (ticker = 'TXN') and (ticker = 'TXT') and (ticker = 'UA')

Failed list (not all) with time.sleep(60):
(ticker = 'BHGE') and (ticker = 'SRCL') and (ticker = 'SYK') and (ticker = 'STI')
